src/install.js

```diff
@@ -50,73 +56,197 @@ export const DEPENDENCY_FIELDS = [
+		})
 	})
 }
 
-export function resolveRemote (nodeModules, parentTarget, name, version) {
-	log(`resolving ${name}@${version} from ${nodeModules} via ${nodeModules}`)
+/**
+ * resolve a dependency's `package.json` file from a remote registry.
+ * @param	{String} nodeModules - `node_modules` base directory.
+ * @param	{String} parentTarget - relative parent's node_modules path.
+ * @param	{String} name - name of the dependency.
+ * @param	{String} version - version of the dependency.
+ * @param	{Boolean} isExplicit - whether the install command asks for an explicit install.
+ * @return {Observable} - observable sequence of `package.json` objects.
+ */
+export function resolveRemote (nodeModules, parentTarget, name, version, isExplicit) {
+	const source = `${name}@${version}`
+	log(`resolving ${source} from remote registry`)
+
+	const parsedSpec = cachedNpa(source)
+
+	switch (parsedSpec.type) {
+		case 'range':
+		case 'version':
+		case 'tag':
+			return resolveFromNpm(nodeModules, parentTarget, parsedSpec)
+		case 'remote':
+			return resolveFromTarball(nodeModules, parentTarget, parsedSpec)
+		case 'hosted':
+			return resolveFromGitHub(nodeModules, parentTarget, parsedSpec)
+		default:
+			throw new Error('Unknown package spec: ' + parsedSpec.type + ' for ' + name)
+	}
+}
 
+/**
+ * resolve a dependency's `package.json` file from the npm registry.
+ * @param	{String} nodeModules - `node_modules` base directory.
+ * @param	{String} parentTarget - relative parent's node_modules path.
+ * @param	{Object} parsedSpec - parsed package name and specifier.
+ * @return {Observable} - observable sequence of `package.json` objects.
+ */
+export function resolveFromNpm (nodeModules, parentTarget, parsedSpec) {
+	const {raw, name, type, spec} = parsedSpec
+	log(`resolving ${raw} from npm`)
 	const options = {...config.httpOptions, retries: config.retries}
-	return registry.match(name, version, options)::map((pkgJson) => {
+	return registry.match(name, spec, options)::map((pkgJson) => {
 		const target = pkgJson.dist.shasum
-		log(`resolved ${name}@${version} to ${target}`)
-		return { parentTarget, pkgJson, target, name, fetch }
+		log(`resolved ${raw} to tarball shasum ${target} from npm`)
+		return { parentTarget, pkgJson, target, name, type, fetch }
+	})
+}
+
+/**
+ * resolve a dependency's `package.json` file from an url tarball.
+ * @param	{String} nodeModules - `node_modules` base directory.
+ * @param	{String} parentTarget - relative parent's node_modules path.
+ * @param	{Object} parsedSpec - parsed package name and specifier.
+ * @return {Observable} - observable sequence of `package.json` objects.
+ */
+export function resolveFromTarball (nodeModules, parentTarget, parsedSpec) {
+	const {raw, name, type, spec} = parsedSpec
+	log(`resolving ${raw} from tarball`)
+	return Observable.create((observer) => {
+		// create shasum from url for storage
+		const hash = crypto.createHash('sha1')
+		hash.update(raw)
```

Comment:
the name / version combination should be alright here. No need to hash the actual package I think.